webapp/channels/src/packages/mattermost-redux/src/utils/data_loader.ts

```diff
@@ -0,0 +1,203 @@
+abstract class DataLoader<Identifier, Result = unknown> {
+    protected readonly fetchBatch: (identifiers: Identifier[]) => Result;
+    private readonly maxBatchSize: number;
+
+    protected readonly pendingIdentifiers = new Set<Identifier>();
+
+    constructor(args: {
+        fetchBatch: (identifiers: Identifier[]) => Result;
+        maxBatchSize: number;
+    }) {
+        this.fetchBatch = args.fetchBatch;
+        this.maxBatchSize = args.maxBatchSize;
+    }
+
+    public queueForLoading(identifiersToLoad: Identifier[]): void {
+        for (const identifier of identifiersToLoad) {
+            if (!identifier) {
+                continue;
+            }
+
+            this.pendingIdentifiers.add(identifier);
+        }
+    }
+
+    /**
+     * startBatch removes an array of identifiers for data to be loaded from pendingIdentifiers and returns it. If
+     * pendingIdentifiers contains more than maxBatchSize identifiers, then only that many are returned, but if it
+     * contains fewer than that, all of the identifiers are returned and pendingIdentifiers is cleared.
+     */
+    protected startBatch(): {identifiers: Identifier[]; moreToLoad: boolean} {
+        let identifiersToLoad;
+
+        // Since we can only fetch a defined number of user statuses at a time, we need to batch the requests
+        if (this.pendingIdentifiers.size >= this.maxBatchSize) {
+            identifiersToLoad = [];
+
+            // We use temp buffer here to store up until max buffer size
+            // and clear out processed user ids
+            for (const identifier of this.pendingIdentifiers) {
+                identifiersToLoad.push(identifier);
+                this.pendingIdentifiers.delete(identifier);
+
+                if (identifiersToLoad.length >= this.maxBatchSize) {
+                    break;
+                }
+            }
+        } else {
+            // If we have less than max buffer size, we can directly fetch the statuses
+            identifiersToLoad = Array.from(this.pendingIdentifiers);
+            this.pendingIdentifiers.clear();
+        }
+
+        return {
+            identifiers: identifiersToLoad,
+            moreToLoad: this.pendingIdentifiers.size > 0,
+        };
+    }
+
+    /**
+     * isBusy is a method for testing which returns true if the DataLoader is waiting to request or receive any data.
+     */
+    public isBusy(): boolean {
+        return this.pendingIdentifiers.size > 0;
+    }
+}
+
+/**
+ * An IntervalDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead
+ * of requesting data immediately, it will periodically check if any objects need to be requested from the server.
+ *
+ * It's intended to be used for loading low priority data such as information needed in response to WebSocket messages
+ * that the user won't see immediately.
+ */
+export class IntervalDataLoader<Identifier, Result = unknown> extends DataLoader<Identifier, Result> {
```

Comment:
It took me a while to understand that this "interval" data loader didn't mean, "repeatedly (re)load the data on a given interval", but more akin to, "load this data once, just be slower about it."

Can we consolidate `IntervalDataLoader` and `DelayedDataLoader`?

Reply:
The naming isn't perfect, but I haven't been able to come up with anything more self explanatory. Maybe `PeriodicDataLoader` could work, but I don't think that's much better.

As for consolidating, I was considering that, but @M-ZubairAhmed had concerns about the performance impact of using repeated timeouts vs a recurring interval when the server and app are under really heavy load. Perhaps once this is in, we can revisit the ceiling tests to see how this would perform in both cases to see if it makes much of a difference

Reply:
Two are equivalent in both terms of server and end result i.e both both them in our case will fire after certain amount of time. I had run the same data loader of previous version with both setTimeout and setInterval and the latter one performed better. We can definetly see the diffrence once we load test to see quantify the difference.

Reply:
Makes sense. Is there a reason to keep the `setTimeout` one at all if we know `setInterval` and fewer allocations are better?

Reply:
I guess its better to quantify difference in both and then we can expire either one of them. At that time when i was working in the PR i just saw setTimeout was a bit poor but how much of that i didnt see and moved on with setInterval. So its better to validate again.

Reply:
In theory, `setTimeout` should be better during periods where there's no data being requested of a given type because there's nothing running while `setInterval` requires a number of running handlers periodically checking for when work is available. I doubt the differences between the two will be very apparent until we use this in a lot of places and we have the server under extreme load. If we find that there are notable differences between the two, perhaps we could even make the `DataLoader` smart enough to swap between them based on how many requests it's getting, but that's obviously way out of scope for an initial version.

We should continue this discussion, but I'm going to move on to QA reviews with this without changing the code, and we can figure out which one to get rid of as a followup. Having two versions is likely redundant, but given they're both written and thoroughly unit tested, reducing that to only one is more of an ideological concern.

I'm definitely very appreciative of this discussion though since it does mean that people really care about this logic and aren't just rubber stamp approving it though :)